### PR TITLE
boards: stm32: add nucleo board stm32g474re

### PR DESCRIPTION
**Pull request by FRASTM**
This patch adds the nucleo_g474re board
     based on the stm32g474re from STMicroelectronics.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

### CODE REVIEW

**Comment by FRASTM** [comment]
add Kconfig.defconfig.stm32g474re with soc

File: dts/arm/st/g4/stm32g474.dtsi [hunk 1]
```diff
@@ -0,0 +1,7 @@
+/*
+ * Copyright (c) 2019 Richard Osterloh <richard.osterloh@gmail.com>
```
Comment (erwango) [inline dts/arm/st/g4/stm32g474.dtsi]: Pleae use your copyright.

File: dts/arm/st/g4/stm32g474Xe.dtsi [hunk 2]
```diff
@@ -0,0 +1,22 @@
+/*
+ * Copyright (c) 2019 Richard Osterloh <richard.osterloh@gmail.com>
```
Comment (erwango) [inline dts/arm/st/g4/stm32g474Xe.dtsi]: Ditto

File: soc/arm/st_stm32/stm32g4/Kconfig.defconfig.stm32g474re [hunk 3]
```diff
@@ -0,0 +1,21 @@
+# Kconfig - STMicroelectronics STM32G474RE MCU
+#
+# Copyright (c) 2019 Richard Osterloh <richard.osterloh@gmail.com>
```
Comment (erwango) [inline soc/arm/st_stm32/stm32g4/Kconfig.defconfig.stm32g474re]: ditto

File: dts/arm/st/g4/stm32g4.dtsi [hunk 4]
```diff
@@ -415,6 +415,18 @@
 			};
 		};
 
+		lptim1: timers@40007c00 {
```
Comment (erwango) [inline dts/arm/st/g4/stm32g4.dtsi]: Please remove for now.

File: dts/arm/st/g4/stm32g474.dtsi [hunk 5]
```diff
@@ -0,0 +1,7 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <st/g4/stm32g4.dtsi>
```
Comment (erwango) [inline dts/arm/st/g4/stm32g474.dtsi]: #include <st/g4/stm32g431.dtsi>

File: soc/arm/st_stm32/stm32g4/Kconfig.defconfig.stm32g474re [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+config SOC
+	string
+	default "stm32g474xx"
+
+config NUM_IRQS
+	int
+	default 102
+
+if GPIO_STM32
```
Comment (erwango) [inline soc/arm/st_stm32/stm32g4/Kconfig.defconfig.stm32g474re]: Remove these lines

File: soc/arm/st_stm32/stm32g4/dts_fixup.h [hunk 7]
```diff
@@ -260,4 +260,11 @@
 #define DT_WWDT_0_CLOCK_BITS			DT_INST_0_ST_STM32_WINDOW_WATCHDOG_CLOCK_BITS
 #define DT_WWDT_0_CLOCK_BUS			DT_INST_0_ST_STM32_WINDOW_WATCHDOG_CLOCK_BUS
 
+#define DT_LPTIM_1_BASE_ADDRESS	DT_ST_STM32_TIMERS_40007C00_BASE_ADDRESS
```
Comment (erwango) [inline soc/arm/st_stm32/stm32g4/dts_fixup.h]: Remove for now.

File: soc/arm/st_stm32/stm32g4/soc.h [hunk 8]
```diff
@@ -78,6 +78,11 @@
 #include <stm32g4xx_ll_pwr.h>
 #endif
 
+#ifdef CONFIG_STM32_LPTIM_TIMER
```
Comment (erwango) [inline soc/arm/st_stm32/stm32g4/soc.h]: Please remove for now

File: boards/arm/nucleo_g431rb/nucleo_g431rb.dts [hunk 9]
```diff
@@ -89,6 +89,10 @@
 	status = "okay";
 };
 
+&lptim1 {
```
Comment (erwango) [inline boards/arm/nucleo_g431rb/nucleo_g431rb.dts]: Please remove for now.

File: boards/arm/nucleo_g431rb/nucleo_g431rb_defconfig [hunk 10]
```diff
@@ -3,7 +3,6 @@
 CONFIG_ARM=y
 CONFIG_SOC_SERIES_STM32G4X=y
 CONFIG_SOC_STM32G431XX=y
-CONFIG_CORTEX_M_SYSTICK=y
```
Comment (erwango) [inline boards/arm/nucleo_g431rb/nucleo_g431rb_defconfig]: Change not related to current PR (different board)

**Review by erwango — CHANGES_REQUESTED** [review]

**Comment by FRASTM** [comment]
change fater review
set stm32g474 flash size to 512K and 128K of SRAM

File: boards/arm/nucleo_g474re/nucleo_g474re_defconfig [hunk 11]
```diff
@@ -0,0 +1,59 @@
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_ARM=y
+CONFIG_SOC_SERIES_STM32G4X=y
+CONFIG_SOC_STM32G474XX=y
+CONFIG_CORTEX_M_SYSTICK=y
```
Comment (erwango) [inline boards/arm/nucleo_g474re/nucleo_g474re_defconfig]: Remove this line

File: dts/arm/st/g4/stm32g474.dtsi [hunk 12]
```diff
@@ -0,0 +1,8 @@
+/*
+ * Copyright (c) 2019 STMicroelectronics.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <st/g4/stm32g4.dtsi>
```
Comment (erwango) [inline dts/arm/st/g4/stm32g474.dtsi]: Remove this line

**Review by erwango — CHANGES_REQUESTED** [review]

File: dts/arm/st/g4/stm32g474.dtsi (hunk 1 above)
Comment (dbkinder) [reply dts/arm/st/g4/stm32g474.dtsi]: Can't just delete Richard's copyright, but you can add Francois'.

File: boards/arm/nucleo_g474re/doc/index.rst [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+
+Nucleo G474RE Board has 6 GPIO controllers. These controllers are responsible for pin muxing,
+input/output, pull-up, etc.
+
+For mode details please refer to `STM32G4 Nucleo-64 board User Manual`_.
+
+Default Zephyr Peripheral Mapping:
+----------------------------------
+
+- UART_1_TX : PC4
```
Comment (dbkinder) [inline boards/arm/nucleo_g474re/doc/index.rst]: This long list will display as a multi-column list if you add this directive and a blank line before the first list item:
```
.. rst-class:: rst-columns

- UART_1_TX : PC4
- UART_1_RX : PC5
  ...
```

**Review by dbkinder — CHANGES_REQUESTED** [review]
doc changes look fine, one suggested edit to make the generated HTML look better.

File: boards/arm/nucleo_g474re/nucleo_g474re_defconfig (hunk 11 above)
Comment (FRASTM) [reply boards/arm/nucleo_g474re/nucleo_g474re_defconfig]: CONFIG_CORTEX_M_SYSTICK=y removed

File: dts/arm/st/g4/stm32g474.dtsi (hunk 12 above)
Comment (FRASTM) [reply dts/arm/st/g4/stm32g474.dtsi]: line "#include <st/g4/stm32g4.dtsi>" replaced by "#include <st/g4/stm32g431.dtsi>"

File: boards/arm/nucleo_g474re/doc/index.rst (hunk 13 above)
Comment (FRASTM) [reply boards/arm/nucleo_g474re/doc/index.rst]: ok, could also fit  nucleo_g431rb/

**Comment by FRASTM** [comment]
change rst file in doc for both boards nucleo_g474re and nucleo_g431rb

**Comment by FRASTM** [comment]
change rst doc for both nucleo_g744re and nucleo_g431rb boards

File: dts/arm/st/g4/stm32g474.dtsi (hunk 12 above)
Comment (erwango) [reply dts/arm/st/g4/stm32g474.dtsi]: I still see both lines actually

File: boards/arm/nucleo_g474re/Kconfig.board [hunk 14]
```diff
@@ -0,0 +1,11 @@
+# Kconfig - STM32G474RE Nucleo board configuration
+#
+# Copyright (c) 2019 Richard Osterloh <richard.osterloh@gmail.com>
```
Comment (erwango) [inline boards/arm/nucleo_g474re/Kconfig.board]: Remove this line

File: boards/arm/nucleo_g474re/Kconfig.defconfig [hunk 15]
```diff
@@ -0,0 +1,45 @@
+# Kconfig - STM32G474RE Nucleo board configuration
+#
+# Copyright (c) 2019 Richard Osterloh <richard.osterloh@gmail.com>
```
Comment (erwango) [inline boards/arm/nucleo_g474re/Kconfig.defconfig]: Remove this line

File: boards/arm/nucleo_g474re/nucleo_g474re_defconfig [hunk 16]
```diff
@@ -3,7 +3,6 @@
 CONFIG_ARM=y
 CONFIG_SOC_SERIES_STM32G4X=y
 CONFIG_SOC_STM32G474XX=y
-CONFIG_CORTEX_M_SYSTICK=y
```
Comment (erwango) [inline boards/arm/nucleo_g474re/nucleo_g474re_defconfig]: Squash this into board commit

File: boards/arm/nucleo_g474re/doc/index.rst [hunk 17]
```diff
@@ -133,6 +133,8 @@ For mode details please refer to `STM32G4 Nucleo-64 board User Manual`_.
 Default Zephyr Peripheral Mapping:
 ----------------------------------
 
+.. rst-class:: rst-columns
```
Comment (erwango) [inline boards/arm/nucleo_g474re/doc/index.rst]: Squash this into board commit

**Review by erwango — CHANGES_REQUESTED** [review]
Please rebase onto latest master as current version can't build (due to missing Cortex Systick activation).

**Comment by FRASTM** [comment]
rebasing on HEAD

**Comment by FRASTM** [comment]
Squash doc rst file  into board commit

**Comment by FRASTM** [comment]
line with wrong Copyright removed

**Comment by FRASTM** [comment]
rebase

**Review by erwango — APPROVED** [review]

**Comment by FRASTM** [comment]
rebase

File: boards/arm/nucleo_g474re/arduino_r3_connector.dtsi [hunk 18]
```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright (c) 2019 STMicroelectronics.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/ {
+	arduino_header: connector {
+		compatible = "arduino-header-r3";
+		#gpio-cells = <2>;
```
Comment (galak) [inline boards/arm/nucleo_g474re/arduino_r3_connector.dtsi]: Need to add `gpio-map-mask` and `gpio-map-pass-thru` properties.

**Review by galak — CHANGES_REQUESTED** [review]
looks good other than the one dts update.

File: boards/arm/nucleo_g474re/arduino_r3_connector.dtsi (hunk 18 above)
Comment (galak) [reply boards/arm/nucleo_g474re/arduino_r3_connector.dtsi]: @erwango and/or @pabigot can help with how these should be set.

Comment (pabigot) [reply boards/arm/nucleo_g474re/arduino_r3_connector.dtsi]:
```
            gpio-map-mask = <0xffffffff 0xffffffc0>;
            gpio-map-pass-thru = <0 0x3f>;
```
for all cases.  See https://docs.zephyrproject.org/latest/guides/dts/index.html#gpio-nexus-nodes

**Review by dbkinder — APPROVED** [review]
LGTM, thanks!

File: boards/arm/nucleo_g474re/arduino_r3_connector.dtsi (hunk 18 above)
Comment (FRASTM) [reply boards/arm/nucleo_g474re/arduino_r3_connector.dtsi]: adding gpio-map-mask and gpio-map-pass-thru properties to boards/arm/nucleo_g474re/arduino_r3_connector.dtsi

**Comment by FRASTM** [comment]
rebase

**Comment by FRASTM** [comment]
rebase

**Review by galak — APPROVED** [review]

**Pull request merged**